File: nsd_android/android/src/main/kotlin/com/haberey/flutter/nsd_android/NsdError.kt
package com.haberey.flutter.nsd_android

import android.net.nsd.NsdManager
import java.lang.Error

internal fun getErrorCause(errorCode: Int): String {
    return when (errorCode) {
        NsdManager.FAILURE_ALREADY_ACTIVE -> ErrorCause.ALREADY_ACTIVE.code
        NsdManager.FAILURE_MAX_LIMIT -> ErrorCause.MAX_LIMIT.code
        else -> ErrorCause.INTERNAL_ERROR.code
    }
}

internal fun getErrorMessage(errorCode: Int): String {
    return when (errorCode) {
        NsdManager.FAILURE_ALREADY_ACTIVE -> "Operation already active"
        NsdManager.FAILURE_MAX_LIMIT -> "Maximum outstanding requests reached"
        else -> "Internal error"
    }
}

/**
 * @param code error cause code as defined by enum in nsd_platform_interface
 */
internal enum class ErrorCause(val code: String) {
    ILLEGAL_ARGUMENT("illegalArgument"),
    ALREADY_ACTIVE("alreadyActive"),
    MAX_LIMIT("maxLimit"),
    INTERNAL_ERROR("internalError"),
    SECURITY_ISSUE("securityIssue"),
}

internal class NsdError(val errorCause: ErrorCause, val errorMessage: String) :
    Error("$errorMessage (${errorCause.code})")

File: nsd_windows/windows/nsd_error.cpp
#include "nsd_error.h"

namespace nsd_windows {

	std::string ToErrorCode(const ErrorCause errorCause)
	{
		switch (errorCause) {
		case ILLEGAL_ARGUMENT:
			return "illegalArgument";

		case ALREADY_ACTIVE:
			return "alreadyActive";

		case MAX_LIMIT:
			return "maxLimit";

		case INTERNAL_ERROR:
		default:
			return "internalError";
		}
	}

	const char* NsdError::what() const throw() {
		return message.c_str(); 
	}

	NsdError::NsdError(const ErrorCause errorCause, const std::string& message) : errorCause(errorCause), message(message)
	{
	}

	NsdError::~NsdError()
	{
	}
}

